launchpad/src/setup.rs

```diff
@@ -204,26 +150,38 @@ pub trait SetupModule {
 
     fn require_valid_time_periods(
         &self,
-        opt_winner_selection_start_epoch: Option<u64>,
         opt_confirm_start_epoch: Option<u64>,
-        opt_confirm_period: Option<u64>,
+        opt_winner_selection_start_epoch: Option<u64>,
         opt_claim_start: Option<u64>,
     ) -> SCResult<()> {
-        let winner_selection_start_epoch = opt_winner_selection_start_epoch
-            .unwrap_or_else(|| self.winner_selection_start_epoch().get());
         let confirm_start_epoch =
             opt_confirm_start_epoch.unwrap_or_else(|| self.confirmation_period_start_epoch().get());
-        let confirm_period =
-            opt_confirm_period.unwrap_or_else(|| self.confirmation_period_in_epochs().get());
+        let winner_selection_start_epoch = opt_winner_selection_start_epoch
+            .unwrap_or_else(|| self.winner_selection_start_epoch().get());
         let claim_start = opt_claim_start.unwrap_or_else(|| self.claim_start_epoch().get());
 
         require!(
-            winner_selection_start_epoch < confirm_start_epoch,
-            "Winner selection start epoch cannot be after confirm start epoch"
+            confirm_start_epoch < winner_selection_start_epoch,
+            "Winner selection start epoch must be after confirm start epoch"
         );
         require!(
-            confirm_start_epoch + confirm_period < claim_start,
-            "Confirm period cannot last over claim period"
+            winner_selection_start_epoch <= claim_start,
+            "Claim period must be after winner selection"
+        );
+
+        Ok(())
+    }
+
+    fn require_launchpad_tokens_deposited(&self) -> SCResult<()> {
+        let amount_per_ticket = self.launchpad_tokens_per_winning_ticket().get();
+        let total_winning_tickets = self.nr_winning_tickets().get();
+        let amount_needed = amount_per_ticket * Self::BigUint::from(total_winning_tickets);
+
+        let launchpad_token_id = self.launchpad_token_id().get();
+        let sc_balance = self.blockchain().get_sc_balance(&launchpad_token_id, 0);
+        require!(
+            sc_balance >= amount_needed,
+            "Wrong launchpad tokens amount deposit by owner or not deposited yet"
```

Comment:
The owner might be allowed to deposit more funds then necessary once users start claiming because `sc_balance` will drop below `amount_needed`. As such, some tokens may be lost permanently if the owner deposits more funds by mistake. The `depositLaunchpadTokens` should be be restricted so that he cannot add more funds after users have started claiming. (`require_add_tickets_period` probably?)